### PR TITLE
Added gumbel_softmax

### PR DESCRIPTION
**Pull request by leonjovanovic**
Issue https://github.com/unifyai/ivy/issues/14420

Changes:

gumbel softmax implemntation for the pytorch frontend
test for gumbel softmax for the pytorch frontend

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py [hunk 1]
```diff
@@ -356,6 +356,26 @@ def softplus(input, beta=1, threshold=20):
...
+def gumbel_softmax(logits, tau=1, hard=False, eps=1e-10, dim=-1):
+    gumbels = -ivy.empty_like(logits).exponential().log()
+    gumbels = (logits + gumbels) / tau
+    y_soft = ivy.softmax(x=gumbels, axis=dim)
+
+    if hard:
+        indices = y_soft.max(axis=dim, keepdims=True)[1]
+        y_hard = ivy.zeros_like(logits).scatter(dim, indices, 1.0)
+        updates = ivy.ones_like(indices)
+        ivy.scatter_nd(indices, updates, reduction="replace", out=y_hard)
```
Comment (Dsantra92) [inline ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py:370]: Any reason as to why the scatter_nd needed here?

File: ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py [hunk 2]
```diff
@@ -356,6 +356,26 @@ def softplus(input, beta=1, threshold=20):
...
+
+        ret = y_hard - y_soft.detach() + y_soft
```
Comment (Dsantra92) [inline ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py:372]: I don't think there are detach functions available for all ivy arrays.

**Review by Dsantra92 — COMMENTED** [review]
Hi @leonjovanovic, thanks for your contribution. I have some minor questions about the implementation. After that, we are good to merge.

File: ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py (hunk 1 above)
Comment (leonjovanovic) [reply ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py]: .scatter(dim, indices, 1.0) was redundant as there is no ivy implementation of that function. I had to either use scatter_flat or scatter_nd but since multidimensionality is required, I chose scatter_nd.

File: ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py (hunk 2 above)
Comment (leonjovanovic) [reply ivy/functional/frontends/torch/nn/functional/non_linear_activation_functions.py]: I replaces .detach() with .stop_gradient(preserve_type=True) which should remove all gradients while working for all ivy.Array

**Comment by Dsantra92** [comment]
LGTM
